### PR TITLE
Adding support for building sources in multiple directories and using  a basic '.dotnetproj' file.

### PR DESCRIPTION
**Pull request by ahsonkhan**

### CODE REVIEW

**Comment by ahsonkhan** [comment]
@KrzysztofCwalina can you please review?

File: demos/dotnet/dotnet/dotnet.cs [hunk 1]
```diff
@@ -27,7 +28,10 @@ static void Main(string[] args)
             Log.IsEnabled = true;
         }
 
-        switch (args[0])
+        var commandArgument = args[0];
```
Comment (KrzysztofCwalina) [inline demos/dotnet/dotnet/dotnet.cs]: At some point we need to rewire this command parsing logic. The current implementation does not really scale past one or two options.

File: demos/dotnet/dotnet/dotnet.cs [hunk 2]
```diff
@@ -43,8 +47,26 @@ static void Main(string[] args)
...
+                else
+                {
+                    Console.WriteLine("Please specify the {0} compiler option.", compilerOptions[0]);
+                }
+                break;
+
+            case "/target":
+                if (compilerOptions.Length > 1)
+                {
+                    Build(args, Log, compilerOptions[1] == "library");
```
Comment (KrzysztofCwalina) [inline demos/dotnet/dotnet/dotnet.cs]: Hmm, could you validate that it's actually exe if we build an exe?

File: demos/dotnet/dotnet/dotnet.cs [hunk 3]
```diff
@@ -64,6 +86,7 @@ static void PrintUsage()
         string appName = Environment.GetCommandLineArgs()[0];
         Console.WriteLine("{0} [/log] - compiles sources in current direcotry. optionally logs diagnostics info.", appName);
         Console.WriteLine("{0} /target:library - compiles sources in current directory into dll.", appName);
+        Console.WriteLine("{0} /recurse:<wildcard> - compiles sources in current directory and subdirectories according to the wildcard specifications.", appName);
```
Comment (KrzysztofCwalina) [inline demos/dotnet/dotnet/dotnet.cs]: Hopefully packages directory will never have cs files :-) And more seriously, I think we need to filter out subdirectories created by the tool: packages, bin, tools.

File: demos/dotnet/dotnet/dotnet.cs [hunk 4]
```diff
@@ -157,8 +180,24 @@ public static ProjectProperties InitializeProperties(string[] args, Log log, boo
         properties.OutputType = buildDll ? ".dll" : ".exe";
         FindCompiler(properties);
 
+        var projectFiles = Directory.GetFiles(properties.ProjectDirectory, "*.dotnetproj");
+        bool projectFileExists = projectFiles.Length > 0;
         // Sources
-        properties.Sources.AddRange(Directory.GetFiles(properties.ProjectDirectory, "*.cs"));
+        if (projectFileExists)
+        {
+            foreach(var projectFile in projectFiles)
```
Comment (KrzysztofCwalina) [inline demos/dotnet/dotnet/dotnet.cs]: Why do you want to enable more than one project file?

File: demos/dotnet/dotnet/dotnet.cs [hunk 5]
```diff
@@ -157,8 +180,24 @@ public static ProjectProperties InitializeProperties(string[] args, Log log, boo
...
+            {
+                properties.Sources.AddRange(ParseProjectFile(properties.ProjectDirectory, projectFile));
```
Comment (KrzysztofCwalina) [inline demos/dotnet/dotnet/dotnet.cs]: I would rather pass properties to the parse method. We will have more things in the project file besides sources (very soon).

File: demos/dotnet/dotnet/dotnet.cs [hunk 6]
```diff
@@ -157,8 +180,24 @@ public static ProjectProperties InitializeProperties(string[] args, Log log, boo
...
+            }
+        }
+        properties.Sources.AddRange(Directory.GetFiles(properties.ProjectDirectory, "*.cs", (args.Length > 1 && args[1] == "*.cs") ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly));
+        if (args.Length > 1 && args[1] != "*.cs")
```
Comment (KrzysztofCwalina) [inline demos/dotnet/dotnet/dotnet.cs]: Why is this indexing just to the second element of the array?

File: demos/dotnet/dotnet/dotnet.cs (hunk 1 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/dotnet.cs]: Agreed. Let's discuss the behavior of the command parsing logic to make it more flexible and scalable.

File: demos/dotnet/dotnet/dotnet.cs (hunk 6 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/dotnet.cs]: Let's say the command argument is generally of the form "/*:*". Once we split the argument and the specification (split over ':'), we have two elements only - the command option and the specification. Once we grab the specification (i.e. whether it is '*.cs' or a subdirectory name), we can get the appropriate source files.

**Comment by KrzysztofCwalina** [comment]
Looks good. Please merge.

**Pull request merged**